File: artifacts/tulipindicators/translate/src/ultosc.rs
use crate::*;

pub fn ti_ultosc_start(options: Option<&[f64]>) -> i32 {
    // Check if options is None (NULL in C) or if it doesn't have at least 3 elements
    match options {
        Some(opts) if opts.len() >= 3 => opts[2] as i32,
        _ => 0, // Default return value when options is None or too short
    }
}
pub fn helper_helper_ti_ultosc_1_1(
    bp_short_sum_ref: &mut f64,
    bp_medium_sum_ref: &mut f64,
    r_short_sum_ref: &mut f64,
    r_medium_sum_ref: &mut f64,
    short_period: i32,
    medium_period: i32,
    long_period: i32,
    bp_buf: &TiBuffer,
    r_buf: &TiBuffer,
    i: i32,
) {
    let mut bp_short_sum = *bp_short_sum_ref;
    let mut bp_medium_sum = *bp_medium_sum_ref;
    let mut r_short_sum = *r_short_sum_ref;
    let mut r_medium_sum = *r_medium_sum_ref;

    let mut short_index = (bp_buf.index - short_period) - 1;
    if short_index < 0 {
        short_index += long_period;
    }
    bp_short_sum -= bp_buf.vals[short_index as usize];
    r_short_sum -= r_buf.vals[short_index as usize];

    if i > medium_period {
        let mut medium_index = (bp_buf.index - medium_period) - 1;
        if medium_index < 0 {
            medium_index += long_period;
        }
        bp_medium_sum -= bp_buf.vals[medium_index as usize];
        r_medium_sum -= r_buf.vals[medium_index as usize];
    }

    *bp_short_sum_ref = bp_short_sum;
    *bp_medium_sum_ref = bp_medium_sum;
    *r_short_sum_ref = r_short_sum;
    *r_medium_sum_ref = r_medium_sum;
}
pub fn helper_ti_ultosc_1(
    output_idx_ref: &mut u32,
    bp_short_sum_ref: &mut f64,
    bp_medium_sum_ref: &mut f64,
    r_short_sum_ref: &mut f64,
    r_medium_sum_ref: &mut f64,
    high: &[f64],
    low: &[f64],
    close: &[f64],
    short_period: i32,
    medium_period: i32,
    long_period: i32,
    output: &mut [f64],
    bp_buf: &mut TiBuffer,
    r_buf: &mut TiBuffer,
    i: i32,
) {
    let mut output_idx = *output_idx_ref;
    let mut bp_short_sum = *bp_short_sum_ref;
    let mut bp_medium_sum = *bp_medium_sum_ref;
    let mut r_short_sum = *r_short_sum_ref;
    let mut r_medium_sum = *r_medium_sum_ref;

    let true_low = if low[i as usize] < close[(i - 1) as usize] {
        low[i as usize]
    } else {
        close[(i - 1) as usize]
    };
    let true_high = if high[i as usize] > close[(i - 1) as usize] {
        high[i as usize]
    } else {
        close[(i - 1) as usize]
    };
    let bp = close[i as usize] - true_low;
    let r = true_high - true_low;

    bp_short_sum += bp;
    bp_medium_sum += bp;
    r_short_sum += r;
    r_medium_sum += r;

    // Process bp_buf
    {
        if bp_buf.pushes >= bp_buf.size {
            bp_buf.sum -= bp_buf.vals[bp_buf.index as usize];
        }
        bp_buf.sum += bp;
        bp_buf.vals[bp_buf.index as usize] = bp;
        bp_buf.pushes += 1;
        bp_buf.index += 1;
        if bp_buf.index >= bp_buf.size {
            bp_buf.index = 0;
        }
    }

    // Process r_buf
    {
        if r_buf.pushes >= r_buf.size {
            r_buf.sum -= r_buf.vals[r_buf.index as usize];
        }
        r_buf.sum += r;
        r_buf.vals[r_buf.index as usize] = r;
        r_buf.pushes += 1;
        r_buf.index += 1;
        if r_buf.index >= r_buf.size {
            r_buf.index = 0;
        }
    }

    if i > short_period {
        helper_helper_ti_ultosc_1_1(
            &mut bp_short_sum,
            &mut bp_medium_sum,
            &mut r_short_sum,
            &mut r_medium_sum,
            short_period,
            medium_period,
            long_period,
            bp_buf,
            r_buf,
            i,
        );
    }

    if i >= long_period {
        let first = (4.0 * bp_short_sum) / r_short_sum;
        let second = (2.0 * bp_medium_sum) / r_medium_sum;
        let third = bp_buf.sum / r_buf.sum;
        let ult = ((first + second + third) * 100.0) / 7.0;
        output[output_idx as usize] = ult;
        output_idx += 1;
    }

    *output_idx_ref = output_idx;
    *bp_short_sum_ref = bp_short_sum;
    *bp_medium_sum_ref = bp_medium_sum;
    *r_short_sum_ref = r_short_sum;
    *r_medium_sum_ref = r_medium_sum;
}
pub fn ti_ultosc(
    size: i32,
    inputs: Option<&[&[f64]]>,
    options: Option<&[f64]>,
    outputs: Option<&mut [&mut [f64]]>,
) -> i32 {
    // Early return if any required input is None
    let inputs = match inputs {
        Some(i) => i,
        None => return 1,
    };
    let options = match options {
        Some(o) => o,
        None => return 1,
    };
    let outputs = match outputs {
        Some(o) => o,
        None => return 1,
    };

    let high = inputs[0];
    let low = inputs[1];
    let close = inputs[2];
    
    let short_period = options[0] as i32;
    let medium_period = options[1] as i32;
    let long_period = options[2] as i32;
    
    let output = &mut outputs[0];
    let mut output_idx = 0;

    // Validation checks
    if short_period < 1 {
        return 1;
    }
    if medium_period < short_period {
        return 1;
    }
    if long_period < medium_period {
        return 1;
    }

    let start = ti_ultosc_start(Some(options));
    if size <= start {
        return 0;
    }

    let mut bp_buf = match ti_buffer_new(long_period) {
        Some(buf) => buf,
        None => return 1,
    };
    let mut r_buf = match ti_buffer_new(long_period) {
        Some(buf) => buf,
        None => return 1,
    };

    let mut bp_short_sum = 0.0;
    let mut bp_medium_sum = 0.0;
    let mut r_short_sum = 0.0;
    let mut r_medium_sum = 0.0;

    for i in 1..size {
        helper_ti_ultosc_1(
            &mut output_idx,
            &mut bp_short_sum,
            &mut bp_medium_sum,
            &mut r_short_sum,
            &mut r_medium_sum,
            high,
            low,
            close,
            short_period,
            medium_period,
            long_period,
            output,
            &mut bp_buf,
            &mut r_buf,
            i,
        );
    }

    ti_buffer_free(*bp_buf);
    ti_buffer_free(*r_buf);

    // Assertion check
    assert!((&output[output_idx as usize] as *const f64 as isize - outputs[0].as_ptr() as isize) / std::mem::size_of::<f64>() as isize == (size - start) as isize);

    0
}

File: artifacts/tulipindicators/decompose/src/ultosc.c
int ti_ultosc_start(const double *options)
{
  return (int) options[2];
}


/*** DEPENDENCIES:
***/


// hint:  ['output_idx_ref is a mutable refrence to unsigned int', 'bp_short_sum_ref is a mutable refrence to double', 'bp_medium_sum_ref is a mutable refrence to double', 'r_short_sum_ref is a mutable refrence to double', 'r_medium_sum_ref is a mutable refrence to double']
void helper_ti_ultosc_1(unsigned int * const output_idx_ref, double * const bp_short_sum_ref, double * const bp_medium_sum_ref, double * const r_short_sum_ref, double * const r_medium_sum_ref, const double * const high, const double * const low, const double * const close, const int short_period, const int medium_period, const int long_period, double * const output, ti_buffer * const bp_buf, ti_buffer * const r_buf, int i)
{
  unsigned int output_idx = *output_idx_ref;
  double bp_short_sum = *bp_short_sum_ref;
  double bp_medium_sum = *bp_medium_sum_ref;
  double r_short_sum = *r_short_sum_ref;
  double r_medium_sum = *r_medium_sum_ref;
  const double true_low = (low[i] < close[i - 1]) ? (low[i]) : (close[i - 1]);
  const double true_high = (high[i] > close[i - 1]) ? (high[i]) : (close[i - 1]);
  const double bp = close[i] - true_low;
  const double r = true_high - true_low;
  bp_short_sum += bp;
  bp_medium_sum += bp;
  r_short_sum += r;
  r_medium_sum += r;
  do
  {
    if (bp_buf->pushes >= bp_buf->size)
    {
      bp_buf->sum -= bp_buf->vals[bp_buf->index];
    }
    bp_buf->sum += bp;
    bp_buf->vals[bp_buf->index] = bp;
    bp_buf->pushes += 1;
    bp_buf->index = bp_buf->index + 1;
    if (bp_buf->index >= bp_buf->size)
    {
      bp_buf->index = 0;
    }
  }
  while (0);
  do
  {
    if (r_buf->pushes >= r_buf->size)
    {
      r_buf->sum -= r_buf->vals[r_buf->index];
    }
    r_buf->sum += r;
    r_buf->vals[r_buf->index] = r;
    r_buf->pushes += 1;
    r_buf->index = r_buf->index + 1;
    if (r_buf->index >= r_buf->size)
    {
      r_buf->index = 0;
    }
  }
  while (0);
  if (i > short_period)
  {
    int short_index = (bp_buf->index - short_period) - 1;
    if (short_index < 0)
    {
      short_index += long_period;
    }
    bp_short_sum -= bp_buf->vals[short_index];
    r_short_sum -= r_buf->vals[short_index];
    if (i > medium_period)
    {
      int medium_index = (bp_buf->index - medium_period) - 1;
      if (medium_index < 0)
      {
        medium_index += long_period;
      }
      bp_medium_sum -= bp_buf->vals[medium_index];
      r_medium_sum -= r_buf->vals[medium_index];
    }
  }
  if (i >= long_period)
  {
    const double first = (4 * bp_short_sum) / r_short_sum;
    const double second = (2 * bp_medium_sum) / r_medium_sum;
    const double third = (1 * bp_buf->sum) / r_buf->sum;
    const double ult = (((first + second) + third) * 100.0) / 7.0;
    output[output_idx] = ult;
    output_idx += 1;
  }
  *output_idx_ref = output_idx;
  *bp_short_sum_ref = bp_short_sum;
  *bp_medium_sum_ref = bp_medium_sum;
  *r_short_sum_ref = r_short_sum;
  *r_medium_sum_ref = r_medium_sum;
}


/*** DEPENDENCIES:
typedef struct 
{
  int size;
  int pushes;
  int index;
  double sum;
  double vals[1];
} ti_buffer
----------------------------
***/


int ti_ultosc(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *high = inputs[0];
  const double *low = inputs[1];
  const double *close = inputs[2];
  const int short_period = (int) options[0];
  const int medium_period = (int) options[1];
  const int long_period = (int) options[2];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (short_period < 1)
  {
    return 1;
  }
  if (medium_period < short_period)
  {
    return 1;
  }
  if (long_period < medium_period)
  {
    return 1;
  }
  if (size <= ti_ultosc_start(options))
  {
    return 0;
  }
  ti_buffer *bp_buf = ti_buffer_new(long_period);
  ti_buffer *r_buf = ti_buffer_new(long_period);
  double bp_short_sum = 0;
  double bp_medium_sum = 0;
  double r_short_sum = 0;
  double r_medium_sum = 0;
  int i;
  for (i = 1; i < size; i += 1)
  {
    helper_ti_ultosc_1(&output_idx, &bp_short_sum, &bp_medium_sum, &r_short_sum, &r_medium_sum, high, low, close, short_period, medium_period, long_period, output, bp_buf, r_buf, i);
  }

  ti_buffer_free(bp_buf);
  ti_buffer_free(r_buf);
  assert(((&output[output_idx]) - outputs[0]) == (size - ti_ultosc_start(options)));
  return 0;
}


/*** DEPENDENCIES:
void ti_buffer_free(ti_buffer *buffer)
{
  free(buffer);
}


----------------------------
int ti_ultosc_start(const double *options)
{
  return (int) options[2];
}


----------------------------
void helper_ti_ultosc_1(unsigned int * const output_idx_ref, double * const bp_short_sum_ref, double * const bp_medium_sum_ref, double * const r_short_sum_ref, double * const r_medium_sum_ref, const double * const high, const double * const low, const double * const close, const int short_period, const int medium_period, const int long_period, double * const output, ti_buffer * const bp_buf, ti_buffer * const r_buf, int i)
{
  unsigned int output_idx = *output_idx_ref;
  double bp_short_sum = *bp_short_sum_ref;
  double bp_medium_sum = *bp_medium_sum_ref;
  double r_short_sum = *r_short_sum_ref;
  double r_medium_sum = *r_medium_sum_ref;
  const double true_low = (low[i] < close[i - 1]) ? (low[i]) : (close[i - 1]);
  const double true_high = (high[i] > close[i - 1]) ? (high[i]) : (close[i - 1]);
  const double bp = close[i] - true_low;
  const double r = true_high - true_low;
  bp_short_sum += bp;
  bp_medium_sum += bp;
  r_short_sum += r;
  r_medium_sum += r;
  do
  {
    if (bp_buf->pushes >= bp_buf->size)
    {
      bp_buf->sum -= bp_buf->vals[bp_buf->index];
    }
    bp_buf->sum += bp;
    bp_buf->vals[bp_buf->index] = bp;
    bp_buf->pushes += 1;
    bp_buf->index = bp_buf->index + 1;
    if (bp_buf->index >= bp_buf->size)
    {
      bp_buf->index = 0;
    }
  }
  while (0);
  do
  {
    if (r_buf->pushes >= r_buf->size)
    {
      r_buf->sum -= r_buf->vals[r_buf->index];
    }
    r_buf->sum += r;
    r_buf->vals[r_buf->index] = r;
    r_buf->pushes += 1;
    r_buf->index = r_buf->index + 1;
    if (r_buf->index >= r_buf->size)
    {
      r_buf->index = 0;
    }
  }
  while (0);
  if (i > short_period)
  {
    int short_index = (bp_buf->index - short_period) - 1;
    if (short_index < 0)
    {
      short_index += long_period;
    }
    bp_short_sum -= bp_buf->vals[short_index];
    r_short_sum -= r_buf->vals[short_index];
    if (i > medium_period)
    {
      int medium_index = (bp_buf->index - medium_period) - 1;
      if (medium_index < 0)
      {
        medium_index += long_period;
      }
      bp_medium_sum -= bp_buf->vals[medium_index];
      r_medium_sum -= r_buf->vals[medium_index];
    }
  }
  if (i >= long_period)
  {
    const double first = (4 * bp_short_sum) / r_short_sum;
    const double second = (2 * bp_medium_sum) / r_medium_sum;
    const double third = (1 * bp_buf->sum) / r_buf->sum;
    const double ult = (((first + second) + third) * 100.0) / 7.0;
    output[output_idx] = ult;
    output_idx += 1;
  }
  *output_idx_ref = output_idx;
  *bp_short_sum_ref = bp_short_sum;
  *bp_medium_sum_ref = bp_medium_sum;
  *r_short_sum_ref = r_short_sum;
  *r_medium_sum_ref = r_medium_sum;
}


----------------------------
ti_buffer *ti_buffer_new(int size)
{
  const int s = ((int) (sizeof(ti_buffer))) + ((size - 1) * ((int) (sizeof(double))));
  ti_buffer *ret = (ti_buffer *) malloc((unsigned int) s);
  ret->size = size;
  ret->pushes = 0;
  ret->index = 0;
  ret->sum = 0;
  return ret;
}


----------------------------
typedef struct 
{
  int size;
  int pushes;
  int index;
  double sum;
  double vals[1];
} ti_buffer
----------------------------
***/


